{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from os import chdir\n",
    "import subprocess\n",
    "import glob\n",
    "import matplotlib.pyplot as plt\n",
    "from functools import reduce\n",
    "\n",
    "# local import (Hirotaka's myfunctions.py script)\n",
    "import myfunctions as myfunc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory 'temp2': File exists\n"
     ]
    }
   ],
   "source": [
    "!mkdir temp2\n",
    "chdir('temp2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the id file, covariate file (all European)\n",
    "wgsfolder = '/data/CARD/PD/GENOMES/august19/genotypes'\n",
    "# wgsfolder = '/data/CARD/PD/WGS/june2019'\n",
    "idfile = '_keep.id'\n",
    "demog = pd.read_csv('/data/LNG/iwakih2/ampPD/data1_out/PP_PD_BF_HB_LC_demog.csv')\n",
    "euro = pd.read_csv(f'/data/CARD/PD/GENOMES/august19/genotypes/PCA_filtered_europeans.txt', sep=' ',\n",
    "                  header=None, names=['FID', 'IID'])\n",
    "df = pd.merge(demog, euro, left_on='WGSID', right_on='IID')\n",
    "df['FID'] = df.WGSID\n",
    "df['IID'] = df.WGSID\n",
    "df.RECRUIT=['CTR' if i=='HC' else i for i in df.RECRUIT]\n",
    "df = df[df.RECRUIT.isin(['CTR', 'PD'])]\n",
    "df = df[df.RECRUIT==df.LASTDIAG]\n",
    "df['pheno'] = [1 if i=='PD' else 0 for i in df.RECRUIT]\n",
    "df['AAD'] = df.AADi.where(df.pheno==0, other=df.AAOi)\n",
    "df['AGE'] = df.AGEatBL.where(df.pheno==0, other=df.AAD)\n",
    "\n",
    "df[['FID', 'IID', 'pheno']].to_csv(idfile, index=False, sep='\\t', header=None)\n",
    "df[['FID', 'IID', 'AGE', 'FEMALE']].to_csv('_qcovar.txt', index=False, sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# used this to generate pca for chr1 just for testing purposes-- prune and merge all chrs first then run PCA\n",
    "with open('_pca.swarm', 'w') as f:\n",
    "    for chrnum in range(1,2):\n",
    "        pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "        out_pcs = f'chr{chrnum}.pca'\n",
    "        plink_pca_cmd = f'\\\n",
    "plink2 --pfile {pfile}\\\n",
    " --pca 280\\\n",
    " --out {out_pcs}\\n'\n",
    "        \n",
    "        f.write(plink_pca_cmd)\n",
    "#         print(plink_pca_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!swarm -f _pca.swarm -g 64 --time=10:00:00 -t 20 --logdir swarm --module plink/2.0_alpha_1_final --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('_script1.swarm', 'w') as f:\n",
    "    for chrnum in range(1,23):\n",
    "#         pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "        pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.freeze9.sqc'\n",
    "        freqfile = f'freq.{chrnum}'\n",
    "        plinkcmd = f\"\\\n",
    "plink2 --pfile {pfile}\\\n",
    " --keep {idfile} \\\n",
    " --freq \\\n",
    " --out {freqfile}\\n\"\n",
    "        f.write(plinkcmd)\n",
    "#         myfunc.shell_do(plinkcmd)\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!swarm -f _script1.swarm -g 64 --time=10:00:00 -t 20 --logdir swarm --module plink/2.0_alpha_1_final --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dask.distributed import Client, progress\n",
    "# client should match the qued number of threads\n",
    "client = Client(processes=False, threads_per_worker=2,\n",
    "                n_workers=2, memory_limit='30GB')\n",
    "import dask\n",
    "import dask.dataframe as dd\n",
    "\n",
    "# testing out just 2 maf bins\n",
    "def MAFbin(x):\n",
    "    if x>0:\n",
    "        if x<0.01:\n",
    "            return(1)\n",
    "        else:\n",
    "            return(2)\n",
    "    else:\n",
    "        return(0)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#try testing fewer maf bins\n",
    "# def MAFbin(x):\n",
    "#     if x>0:\n",
    "#         if x<0.001:\n",
    "#             return(1)\n",
    "#         elif x<0.01:\n",
    "#             return(2)\n",
    "#         elif x<0.1:\n",
    "#             return(3)\n",
    "#         elif x<0.25:\n",
    "#             return(4)\n",
    "#         elif x<0.5:\n",
    "#             return(5)\n",
    "#         else:\n",
    "#             return(6)\n",
    "#     else:\n",
    "#         return(0)\n",
    "\n",
    "\n",
    "\n",
    "# def MAFbin(x):\n",
    "#     if x>0:\n",
    "#         if x<0.001:\n",
    "#             return(1)\n",
    "#         elif x<0.01:\n",
    "#             return(2)\n",
    "#         elif x<0.1:\n",
    "#             return(3)\n",
    "#         elif x<0.2:\n",
    "#             return(4)\n",
    "#         elif x<0.3:\n",
    "#             return(5)\n",
    "#         elif x<0.4:\n",
    "#             return(6)\n",
    "#         else:\n",
    "#             return(7)\n",
    "#     else:\n",
    "#         return(0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "client.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# total_counts_df = pd.DataFrame(index=[0,1,2,3,4,5,6])\n",
    "#test for 2 maf bins\n",
    "total_counts_df = pd.DataFrame(index=[0,1,2])\n",
    "\n",
    "for chrnum in range(1,23):\n",
    "    freqfile = f'freq.{chrnum}'\n",
    "\n",
    "    df = dd.read_csv(f'{freqfile}.afreq', sep='\\t')\n",
    "    df['MAF'] = df.ALT_FREQS.where(cond=df.ALT_FREQS<=0.5, other=1-df.ALT_FREQS)\n",
    "    df['Group'] = df.MAF.apply(MAFbin, meta=('MAF', 'float64'))\n",
    "    for i in range(1,3):\n",
    "        df.loc[df.Group==i, ['ID']].to_csv(\n",
    "            f'{freqfile}.mafbin{i}', index=False, single_file=True)\n",
    "    print(f'chr{chrnum} finished')\n",
    "    total_counts_df[chrnum] = pd.DataFrame(df.groupby('Group').ID.count().compute())\n",
    "#     total_counts_df.merge(counts_df, left_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add chr names\n",
    "chrs = ['chr' + str(i) for i in range(1,23)]\n",
    "total_counts_df.columns = chrs\n",
    "\n",
    "# add bins\n",
    "# bins = ['0.000', '<0.001', '<0.01', '<0.1', '<0.25', '<0.5', '=0.5']\n",
    "bins = ['0.000', '<0.01', '>=0.01']\n",
    "total_counts_df['bin'] = bins\n",
    "total_counts_df.set_index('bin', inplace=True)\n",
    "total_counts_df['total'] = total_counts_df.sum(axis=1)\n",
    "total_counts_df.to_csv('_maf_bin_counts.txt', sep='\\t')\n",
    "total_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# freqfile = f'freq.21'\n",
    "\n",
    "# df = pd.read_csv(f'{freqfile}.afreq', sep='\\t')\n",
    "# df['MAF'] = df.ALT_FREQS.where(cond=df.ALT_FREQS<=0.5, other=1-df.ALT_FREQS)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df['Group'] = df.MAF.apply(MAFbin)\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df[df.Group == 6].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('_script2.swarm', 'w') as f:\n",
    "    for chrnum in range(1,23):\n",
    "        pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "        freqfile = f'freq.{chrnum}'\n",
    "        for Indx in range(1,3):\n",
    "            mafbinIndx = f'mafbin{Indx}'\n",
    "            mafbinfile =  f'{freqfile}.{mafbinIndx}'\n",
    "\n",
    "            plinkcmd = f\"\\\n",
    "plink2 --pfile {pfile}\\\n",
    " --extract {mafbinfile}\\\n",
    " --keep {idfile}\\\n",
    " --make-bed\\\n",
    " --out {mafbinfile}\"\n",
    "            gctacmd1 = f'\\\n",
    "gcta64 --bfile {mafbinfile}\\\n",
    " --ld-score-region 10000\\\n",
    " --out {mafbinfile}.ldbin\\\n",
    " --thread-num 10'\n",
    "\n",
    "            f.write(f'{plinkcmd} && {gctacmd1}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!swarm -f _script2.swarm --time=40:00:00 -g 64 -t 64 --logdir swarm --module GCTA,plink/2.0_alpha_1_final --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to rerun for jobs that were killed\n",
    "outputs = glob.glob('freq.*.mafbin*.ldbin.score.ld')\n",
    "# len(glob.glob('*.ldbin.log'))\n",
    "expected_out = ['freq.' + str(i) + '.mafbin' + str(j) + '.ldbin.score.ld' for i in range(1,23) for j in range(1,3)]\n",
    "missing = [bed.replace('.ldbin.score.ld', '') for bed in list(set(expected_out) - set(outputs))]\n",
    "\n",
    "with open('_rerunscript.swarm', 'w') as f:\n",
    "    for mafbinfile in missing:\n",
    "        gctacmd1 = f'\\\n",
    "gcta64 --bfile {mafbinfile}\\\n",
    " --ld-score-region 10000\\\n",
    " --out {mafbinfile}.ldbin\\\n",
    " --thread-num 10'\n",
    "\n",
    "        f.write(f'{gctacmd1}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !swarm -f _rerunscript.swarm --time=40:00:00 -g 64 -t 10 --logdir swarm --module GCTA --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for chrnum in range(1,23):\n",
    "    pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "    freqfile = f'freq.{chrnum}'\n",
    "    for Indx in range(1,3):\n",
    "        mafbinIndx = f'mafbin{Indx}'\n",
    "        mafbinfile =  f'{freqfile}.{mafbinIndx}'\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "        # Note this is the ldscore within the maf bin\n",
    "        df = dd.read_csv(f'{mafbinfile}.ldbin.score.ld', sep=' ')\n",
    "        ldscvec = df.ldscore_SNP.compute()\n",
    "        ldscThres = ldscvec.median()\n",
    "        df['Group'] = 1\n",
    "        df['Group'] = df.Group.where(cond=df.ldscore_SNP<ldscThres, other=2)\n",
    "\n",
    "\n",
    "        # Create ldbin and GRMs within the ldbin\n",
    "        for i in range(1, 3):\n",
    "            ldbinIndx = f'{mafbinfile}.ldbin{i}'\n",
    "            df.loc[df.Group==i, ['SNP']].to_csv(\n",
    "                f'{ldbinIndx}', index=False, single_file=True)\n",
    "            gctacmd2 = f'\\\n",
    "            gcta64 --bfile {mafbinfile}\\\n",
    " --extract {ldbinIndx}\\\n",
    " --make-grm\\\n",
    " --out {ldbinIndx}'\n",
    "            myfunc.shell_do(gctacmd2)\n",
    "            with open('multi_GRMs.txt', 'a') as f:\n",
    "                f.write(f'{ldbinIndx}\\n')\n",
    "        # mafbin information save to mafbinInfo.csv\n",
    "        with open('mafbinInfo.txt', 'a') as f:\n",
    "            s = f'{chrnum}\\t{mafbinIndx}\\t{len(ldscvec)}\\t{ldscvec.mean()}\\t{ldscvec.median()}\\t{ldscvec.std()}'\n",
    "            f.write(f'{s}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create covar file per chrom and impute missing age (may want to consider dropping age column in this case (missing for more than half))\n",
    "# impute \"FEMALE\" with median\n",
    "# for chrnum in range(1,2):\n",
    "#     pcs_df = pd.read_csv(f'chr{chrnum}.pca.eigenvec', sep='\\t')\n",
    "#     pcs_df.rename(columns={'#FID':'FID'}, inplace=True)\n",
    "#     covar_df = df[['FID', 'IID', 'AGE', 'FEMALE']].merge(pcs_df, on=['FID','IID'])\n",
    "#     covar_df.drop(['AGE','FEMALE'], axis=1, inplace=True)\n",
    "# #     covar_df.AGE.fillna(covar_df.AGE.mean(), inplace=True)\n",
    "# #     covar_df.FEMALE.fillna(covar_df.FEMALE.median(),inplace=True)\n",
    "# #     print(f'dropping {covar_df[covar_df.AGE.isnull()].shape[0]} rows for missing age')\n",
    "# #     covar_df.dropna(how='any', inplace=True)\n",
    "#     covar_df.to_csv(f'chr{chrnum}.covar', index=False, sep=' ', header=None)\n",
    "#     print(f'wrote: chr{chrnum}.covar with {covar_df.shape[0]} samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# covar_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# count in each ld bin\n",
    "high_ld_counts_dict = {i:0 for i in range(1,3)}\n",
    "low_ld_counts_dict = {i:0 for i in range(1,3)}\n",
    "\n",
    "for chrnum in range(1,23):\n",
    "    freqfile = f'freq.{chrnum}'\n",
    "    for Indx in range(1,3):\n",
    "        mafbinIndx = f'mafbin{Indx}'\n",
    "        mafbinfile =  f'{freqfile}.{mafbinIndx}'\n",
    "        for i in range(1, 3):\n",
    "            ldbinIndx = f'{mafbinfile}.ldbin{i}'\n",
    "            ldbin_df = pd.read_csv(ldbinIndx,sep='\\t')\n",
    "            if (i == 1):\n",
    "                low_ld_counts_dict[Indx] += ldbin_df.shape[0]\n",
    "            else:\n",
    "                high_ld_counts_dict[Indx] += ldbin_df.shape[0]\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'high ld counts per maf group: {high_ld_counts_dict}')\n",
    "print(f'low ld counts per maf group: {low_ld_counts_dict}')\n",
    "high_df = pd.Series(high_ld_counts_dict).to_frame()\n",
    "low_df = pd.Series(low_ld_counts_dict).to_frame()\n",
    "high_df.columns = ['high_ld']\n",
    "low_df.columns = ['low_ld']\n",
    "ld_counts_df = high_df.merge(low_df, left_index=True, right_index=True)\n",
    "# bins = ['<0.001', '<0.01', '<0.1', '<0.25', '<0.5', '=0.5']\n",
    "bins = ['<0.01','>=0.01']\n",
    "ld_counts_df['maf_bin'] = bins\n",
    "ld_counts_df.set_index('maf_bin', inplace=True)\n",
    "ld_counts_df.to_csv('_ld_bin_counts.txt', sep='\\t')\n",
    "ld_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --reml-no-lrt\\\n",
    "# --reml-inv-mtd 3\\\n",
    "# run with all maf bins\n",
    "# gctacmd3 = f'\\\n",
    "# gcta --reml\\\n",
    "#  --mgrm multi_GRMs.txt\\\n",
    "#  --pheno {idfile}\\\n",
    "#  --qcovar chr1.pca.10.eigenvec\\\n",
    "#  --reml-alg 2\\\n",
    "#  --reml-no-constrain\\\n",
    "#  --thread-num 64\\\n",
    "#  --reml-maxit 1000\\\n",
    "#  --out test'\n",
    "# # myfunc.shell_do(gctacmd3)\n",
    "\n",
    "# with open('_reml_script.swarm', 'w') as f:\n",
    "# #     f.write('#!/bin/bash\\n')\n",
    "#     f.write(f'{gctacmd3}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !swarm -f _reml_script.swarm --time=80:00:00 -g 64 -t 64 --logdir swarm --module GCTA --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #set up to run with just maf bins 2-6 as we might not have resolution for maf<0.001\n",
    "# for chrnum in range(1,23):\n",
    "#     pfile = f'{wgsfolder}/pd.june2019.chr{chrnum}.sqc'\n",
    "#     freqfile = f'freq.{chrnum}'\n",
    "#     for Indx in range(2,6): # leaving last maf bin out for testing!!!! ( will be combined with 0.25<maf<=0.5)\n",
    "#         mafbinIndx = f'mafbin{Indx}'\n",
    "#         mafbinfile =  f'{freqfile}.{mafbinIndx}'\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "#         # Note this is the ldscore within the maf bin\n",
    "#         df = dd.read_csv(f'{mafbinfile}.ldbin.score.ld', sep=' ')\n",
    "#         ldscvec = df.ldscore_SNP.compute()\n",
    "#         ldscThres = ldscvec.median()\n",
    "#         df['Group'] = 1\n",
    "#         df['Group'] = df.Group.where(cond=df.ldscore_SNP<ldscThres, other=2)\n",
    "\n",
    "\n",
    "#         # Create ldbin and GRMs within the ldbin\n",
    "#         for i in range(1, 3):\n",
    "#             ldbinIndx = f'{mafbinfile}.ldbin{i}'\n",
    "\n",
    "#             with open('2_5_maf_bins_multi_GRMs.txt', 'a') as f:\n",
    "#                 f.write(f'{ldbinIndx}\\n')\n",
    "#         # mafbin information save to mafbinInfo.csv\n",
    "#         with open('2_5_mafbinInfo.txt', 'a') as f:\n",
    "#             s = f'{chrnum}\\t{mafbinIndx}\\t{len(ldscvec)}\\t{ldscvec.mean()}\\t{ldscvec.median()}\\t{ldscvec.std()}'\n",
    "#             f.write(f'{s}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gctacmd_2_5_maf = f'\\\n",
    "# gcta --reml\\\n",
    "#  --mgrm 2_5_maf_bins_multi_GRMs.txt\\\n",
    "#  --pheno {idfile}\\\n",
    "#  --qcovar chr1.pca.eigenvec\\\n",
    "#  --reml-alg 2\\\n",
    "#  --reml-no-constrain\\\n",
    "#  --thread-num 64\\\n",
    "#  --reml-maxit 1000\\\n",
    "#  --out test_2_5_maf'\n",
    "# # myfunc.shell_do(gctacmd3)\n",
    "\n",
    "# with open('2_5_maf_bins_reml_script.swarm', 'w') as f:\n",
    "# #     f.write('#!/bin/bash\\n')\n",
    "#     f.write(f'{gctacmd_2_5_maf}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !swarm -f 2_5_maf_bins_reml_script.swarm --time=80:00:00 -g 64 -t 64 --logdir swarm --module GCTA --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get estimates for cpu-hrs for analysis\n",
    "jobs_df = pd.read_csv('../biowulf_user_dashboard.csv')\n",
    "jobs_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "jobs_df2 = jobs_df[jobs_df.state == 'COMPLETED'][['jobid','cpu', 'cpu_min',\n",
    "       'cpu_max', 'cpu_avg', 'cpu_util', 'mem', 'mem_min', 'mem_max','elapsed_time']]\n",
    "\n",
    "jobs_df2['hrs'] = jobs_df2.elapsed_time/3600\n",
    "jobs_df2['cpu_hrs'] = jobs_df2.cpu * jobs_df2.hrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(jobs_df2.cpu_hrs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now try merging grms\n",
    "\n",
    "# !gcta --mgrm multi_GRMs.txt --make-grm --out test_merged_grm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gctacmd_merged = f'\\\n",
    "# gcta --reml\\\n",
    "#  --grm test_merged_grm\\\n",
    "#  --pheno {idfile}\\\n",
    "#  --qcovar chr1.pca.eigenvec\\\n",
    "#  --reml-alg 2\\\n",
    "#  --reml-no-constrain\\\n",
    "#  --thread-num 64\\\n",
    "#  --reml-maxit 1000\\\n",
    "#  --out test_merged'\n",
    "\n",
    "# with open('merged_reml_script.swarm', 'w') as f:\n",
    "# #     f.write('#!/bin/bash\\n')\n",
    "#     f.write(f'{gctacmd_merged}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !swarm -f merged_reml_script.swarm --time=80:00:00 -g 64 -t 64 --logdir swarm --module GCTA --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep_ids = pd.read_csv('_keep.id',sep='\\t',header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep_ids[2].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now merge chr grms (1 grm per mafbin/ldbin combo: 6 maf bins, 2 ldbins = 12 total grms)\n",
    "\n",
    "with open('multi_GRMs.merged_chrs.txt', 'w') as mergedGRMlist:\n",
    "    for Indx in range(1,3): \n",
    "        mafbinIndx = f'mafbin{Indx}'\n",
    "        for i in range(1, 3):\n",
    "            ldbinIndx = f'{mafbinIndx}.ldbin{i}'\n",
    "            mergedGRMfile = f'freq.{ldbinIndx}'\n",
    "            for chrnum in range(1,23):\n",
    "                freqfile = f'freq.{chrnum}.{ldbinIndx}'\n",
    "                multiGRMfile = f'multi_GRMs.{ldbinIndx}.txt'\n",
    "                with open(f'{multiGRMfile}', 'a') as f:\n",
    "                    f.write(f'{freqfile}\\n')\n",
    "            \n",
    "            gcta_merge_cmd = f'\\\n",
    "gcta --mgrm {multiGRMfile}\\\n",
    " --make-grm --out {mergedGRMfile}'\n",
    "            myfunc.shell_do(gcta_merge_cmd)\n",
    "            mergedGRMlist.write(f'{mergedGRMfile}\\n')       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #  \n",
    "# gctacmd_merged_chrs_reml = f'\\\n",
    "# gcta --reml\\\n",
    "#  --mgrm multi_GRMs.merged_chrs.txt\\\n",
    "#  --pheno {idfile}\\\n",
    "#  --qcovar chr1.pca.10.eigenvec\\\n",
    "#  --reml-alg 1\\\n",
    "#  --reml-no-constrain\\\n",
    "#  --thread-num 64\\\n",
    "#  --reml-maxit 1000\\\n",
    "#  --out test_alg1'\n",
    "\n",
    "# with open('_reml_alg1_script.swarm', 'w') as f:\n",
    "#     f.write(f'{gctacmd_merged_chrs_reml}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !swarm -f _reml_alg1_script.swarm --time=80:00:00 -g 64 -t 64 --logdir swarm --module GCTA --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gctacmd_merged_chrs_reml = f'\\\n",
    "# gcta --reml\\\n",
    "#  --mgrm multi_GRMs.merged_chrs.txt\\\n",
    "#  --pheno {idfile}\\\n",
    "#  --qcovar chr1.pca.10.eigenvec\\\n",
    "#  --reml-alg 2\\\n",
    "#  --reml-no-constrain\\\n",
    "#  --thread-num 64\\\n",
    "#  --reml-maxit 1000\\\n",
    "#  --out test_alg2'\n",
    "\n",
    "# with open('_reml_alg2_script.swarm', 'w') as f:\n",
    "#     f.write(f'{gctacmd_merged_chrs_reml}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !swarm -f _reml_alg2_script.swarm --time=80:00:00 -g 64 -t 64 --logdir swarm --module GCTA --partition=norm --devel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now fire off jobs in a loop like gridsearch\n",
    "# \n",
    "param_grid = {'--reml-inv-mtd': [None, 3, 4], \n",
    "              '--reml-alg': [0, 1, 2]\n",
    "             }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reml_gridsearch_cmds = []\n",
    "for opts in itertools.product(param_grid['--reml-alg'], param_grid['--reml-inv-mtd']):\n",
    "    if opts[1]:\n",
    "        gctacmd = f'\\\n",
    "gcta --reml\\\n",
    " --mgrm multi_GRMs.merged_chrs.txt\\\n",
    " --pheno {idfile}\\\n",
    " --qcovar chr1.pca.10.eigenvec\\\n",
    " --reml-alg {opts[0]}\\\n",
    " --reml-no-constrain\\\n",
    " --reml-inv-mtd {opts[1]}\\\n",
    " --reml-no-lrt\\\n",
    " --thread-num 64\\\n",
    " --reml-maxit 1000\\\n",
    " --out reml/test_alg{opts[0]}_invmtd{opts[1]}_lrt'\n",
    "        reml_gridsearch_cmds.append(gctacmd) \n",
    "    else:\n",
    "        gctacmd = f'\\\n",
    "gcta --reml\\\n",
    " --mgrm multi_GRMs.merged_chrs.txt\\\n",
    " --pheno {idfile}\\\n",
    " --qcovar chr1.pca.10.eigenvec\\\n",
    " --reml-alg {opts[0]}\\\n",
    " --reml-no-constrain\\\n",
    " --thread-num 64\\\n",
    " --reml-maxit 1000\\\n",
    " --out reml/test_alg{opts[0]}'\n",
    "        reml_gridsearch_cmds.append(gctacmd)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('reml_gridsearch.swarm', 'w') as f:\n",
    "    for cmd in reml_gridsearch_cmds:\n",
    "        f.write(f'{cmd}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!swarm -f reml_gridsearch.swarm --time=80:00:00 -g 64 -t 64 --logdir swarm --module GCTA --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "\n",
    "with open('multi_pcs_reml.swarm', 'w') as f:\n",
    "    for i in [20,40,80]:\n",
    "        \n",
    "        remlcmd = f'\\\n",
    "gcta --reml\\\n",
    " --mgrm multi_GRMs.merged_chrs.txt\\\n",
    " --pheno _keep.id\\\n",
    " --qcovar chr1.pca.{i}.eigenvec\\\n",
    " --reml-alg 0\\\n",
    " --reml-no-constrain\\\n",
    " --thread-num 64\\\n",
    " --reml-maxit 1000\\\n",
    " --out reml/reml_2maf_alg0_{i}pcs'\n",
    "        \n",
    "        f.write(f'{remlcmd}\\n')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "62368810\n"
     ]
    }
   ],
   "source": [
    "!swarm -f multi_pcs_reml.swarm --time=80:00:00 -g 64 -t 64 --logdir swarm --module GCTA --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>Variance20</th>\n",
       "      <th>SE20</th>\n",
       "      <th>Variance40</th>\n",
       "      <th>SE40</th>\n",
       "      <th>Variance80</th>\n",
       "      <th>SE80</th>\n",
       "      <th>maf_ld_group</th>\n",
       "      <th>maf</th>\n",
       "      <th>ld</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>V(G1)/Vp</td>\n",
       "      <td>-1.869264</td>\n",
       "      <td>0.116384</td>\n",
       "      <td>-1.861665</td>\n",
       "      <td>0.117298</td>\n",
       "      <td>-1.832360</td>\n",
       "      <td>0.119221</td>\n",
       "      <td>mafbin1_ldbin1</td>\n",
       "      <td>low</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>V(G2)/Vp</td>\n",
       "      <td>-0.314484</td>\n",
       "      <td>0.031933</td>\n",
       "      <td>-0.312578</td>\n",
       "      <td>0.032534</td>\n",
       "      <td>-0.314913</td>\n",
       "      <td>0.032919</td>\n",
       "      <td>mafbin1_ldbin2</td>\n",
       "      <td>low</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V(G3)/Vp</td>\n",
       "      <td>2.422970</td>\n",
       "      <td>0.136428</td>\n",
       "      <td>2.414762</td>\n",
       "      <td>0.137823</td>\n",
       "      <td>2.426019</td>\n",
       "      <td>0.139473</td>\n",
       "      <td>mafbin2_ldbin1</td>\n",
       "      <td>high</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V(G4)/Vp</td>\n",
       "      <td>-0.080546</td>\n",
       "      <td>0.059133</td>\n",
       "      <td>-0.081502</td>\n",
       "      <td>0.060004</td>\n",
       "      <td>-0.092989</td>\n",
       "      <td>0.060118</td>\n",
       "      <td>mafbin2_ldbin2</td>\n",
       "      <td>high</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sum of V(G)/Vp</td>\n",
       "      <td>0.158676</td>\n",
       "      <td>0.140158</td>\n",
       "      <td>0.159017</td>\n",
       "      <td>0.140744</td>\n",
       "      <td>0.185757</td>\n",
       "      <td>0.141623</td>\n",
       "      <td>total</td>\n",
       "      <td>total</td>\n",
       "      <td>total</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Source  Variance20      SE20  Variance40      SE40  Variance80  \\\n",
       "0        V(G1)/Vp   -1.869264  0.116384   -1.861665  0.117298   -1.832360   \n",
       "1        V(G2)/Vp   -0.314484  0.031933   -0.312578  0.032534   -0.314913   \n",
       "2        V(G3)/Vp    2.422970  0.136428    2.414762  0.137823    2.426019   \n",
       "3        V(G4)/Vp   -0.080546  0.059133   -0.081502  0.060004   -0.092989   \n",
       "4  Sum of V(G)/Vp    0.158676  0.140158    0.159017  0.140744    0.185757   \n",
       "\n",
       "       SE80    maf_ld_group    maf     ld  \n",
       "0  0.119221  mafbin1_ldbin1    low    low  \n",
       "1  0.032919  mafbin1_ldbin2    low   high  \n",
       "2  0.139473  mafbin2_ldbin1   high    low  \n",
       "3  0.060118  mafbin2_ldbin2   high   high  \n",
       "4  0.141623           total  total  total  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# plotting H^2\n",
    "dfs = []\n",
    "for i in [20,40,80]:\n",
    "    hsq = pd.read_csv(f\"reml/reml_2maf_alg0_{i}pcs.hsq\", sep=\"\\t\")\n",
    "    hsq.columns = [\"Source\", f\"Variance{i}\", f\"SE{i}\"]\n",
    "    dfs.append(hsq)\n",
    "# hsq20 = pd.read_csv(\"reml/reml_2maf_alg0_20pcs.hsq\", sep=\"\\t\")\n",
    "# hsq40 = pd.read_csv(\"reml/reml_2maf_alg0_40pcs.hsq\", sep=\"\\t\")\n",
    "# hsq80 = pd.read_csv(\"reml/reml_2maf_alg0_80pcs.hsq\", sep=\"\\t\")\n",
    "\n",
    "# dfs = [hsq20, hsq40, hsq80]\n",
    "hsq = reduce(lambda left,right: pd.merge(left,right,on='Source'), dfs)\n",
    "hsq2 = hsq.iloc[6:11,:].reset_index().drop('index',axis=1)\n",
    "hsq2['maf_ld_group'] = ['mafbin1_ldbin1','mafbin1_ldbin2','mafbin2_ldbin1','mafbin2_ldbin2','total']\n",
    "hsq2['maf'] = ['low','low','high','high','total']\n",
    "# ['0.0001-0.01', '0.0001-0.01', '0.01-0.5', '0.01-0.5','total']\n",
    "hsq2['ld'] = ['low','high','low','high','total']\n",
    "hsq2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "hsq_var = hsq2[[\"Variance20\",\"Variance40\",\"Variance80\",\"maf\",\"ld\"]]\n",
    "hsq_var_melt = pd.melt(hsq_var, id_vars=['maf','ld'],value_vars=['Variance20','Variance40','Variance80'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3cAAAHwCAYAAAAFGZZfAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3dfbSdZX0n/O9FCAMhNo0RpkAgsWIlpHkxCRhHC9GAUGopdHgMWXSBylNc0EqttNKCTLVjB1eZxYPpCx0sPGhJDW2mmbQOSiUmogE65gRShUDh0aRJwBBCCYqAJlzPHzmcEjzk9Zyzz7nO57PWWe7X6/7tnS2/67uve993qbUGAACAoe2gThcAAADAgRPuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAHVVKebCUMmcvH7uulHJaP5cEQ5JwB31oIBtOKWVOKWXja9x3aynlR6WU73f/fbuUcm0pZcxA1AYAL+utN5ZS3l9K+cbL12utk2utK/ph27ts51X3rSilvNDdJ58tpXSVUn6vlPIf+roOGCjCHbTrj2utr0tyRJIPJJmdZGUp5fDOlgUAg8ZvdvfKo5JckeT8JHeUUkpny4L9I9zBACml/Hop5bFSytOllL8vpRzdffsnSyl/0n15ZCnluVLKH3dfP6z7W8Wx+7vdWusLtdZvJjk7ybjsDHoAMGi8cnWvu/d9rpTyb6WUtaWUj/Wyp8r0Uso/l1K2lVJuL6UceiDbr7U+171yeHaStyf5pQMZDzpFuIMBUEp5d5Jrk7wvO78dXJ9kUffdX0syp/vySUm+l+TU7utvT/JIrfXfDrSGWuv3k3wlyS8c6FgA0I/+IMnEJD+b5PQkv9bLY96X5Mwkb0wyNcn7+2LDtdZ/TbIqeiVDlHAHA+OCJLfUWlfXWl9M8vtJ3l5KmZjk3iRvLqWMS3JKkpuTHFNKGZ2dIe9rfVjH40le34fjAcDe+F+llGde/kvy57t57PuS/Lda67/VWjcmWdDLYxbUWh+vtT6d5B+STO/DWvVKhizhDgbG0dm5WpckqbX+IMnWJMfUWp/Pzm8JT83OcPe1JPckeUf6Ptwdk+TpPhwPAPbGObXWn375L8llu3ns0Uk2vOL6hl4e871XXP5hktF9UOPL9EqGLOEOBsbjSSa8fKX7oCbjkmzqvulrSd6d5K1Jvtl9/YwkJye5uy8K6F4JPC3J1/tiPADoJ08kGf+K68cO1IZLKccmmRm9kiFKuIO+N7KUcugr/g5O8tdJPlBKmd59iOX/luSfaq3rup/ztSQXJnmo1vqjJCuS/N9Jvltr3bK7jb1qW4e++ghfpZT/UEqZmeR/Jfm3JP9vH75WAOhrf5Pk90spY0spxyT5zQMcr7y6V/bygFGllFOTLE3yf5LccYDbhI4Q7qDv3ZHk+Vf8faLWuizJNUn+Z3Z+I/mm7Dzc8svuSXJY/n2V7qEkL2TPq3bHvGpbz3ePnSQfK6V8Pzt3Lfl8kq4k/6nW+tyBvDgA6Gd/mGRjku8muSvJ4iQvHsB4/ymv6pXdX7wmyZ9298rNSW7Izj59Zq31pQPYHnRMqbV2ugYAAOhVKeXSJOfXWk/d44NhmLNyBwDAoFFKOaqU8o5SykGllLdk58nFl3S6LhgKDt7zQwAAYMAckuR/ZOc57J7JzvPC7u7UCUA3u2UCAAA0wG6ZAAAADRDuAAAAGjCkfnP3hje8oU6cOLHTZQDQz7q6up6qtR7R6TqGCv0RYPjYXY8cUuFu4sSJWbVqVafLAKCflVLWd7qGoUR/BBg+dtcj7ZYJAADQAOEOAACgAcIdAABAA4bUb+568+Mf/zgbN27MCy+80OlShoVDDz0048ePz8iRIztdCgAAA8i8e2Dtz7x7yIe7jRs35nWve10mTpyYUkqny2larTVbt27Nxo0b88Y3vrHT5QAAMIDMuwfO/s67h/xumS+88ELGjRvnAzYASikZN26cb2sAAIYh8+6Bs7/z7iEf7pL4gA0g7zUAwPBlLjhw9ue9biLcDQVnnXVWnnnmmd0+ZvTo0b3e/v73vz+LFy/uj7IAAKA5w3XuPeR/czfY1VpTa80dd9zR6VIAAKBpw33ubeVuL1155ZX58z//857rn/jEJ/LJT34yc+fOzYwZMzJlypQsXbo0SbJu3bpMmjQpl112WWbMmJENGzZk4sSJeeqpp5Ik55xzTmbOnJnJkyfnpptu2mU7V1xxRWbMmJG5c+dmy5YtP1FHV1dXTj311MycOTNnnHFGnnjiiX581QAAMPDMvffTy+l2KPzNnDmzvtpDDz30E7f1h9WrV9dTTjml5/qkSZPq+vXr67Zt22qttW7ZsqW+6U1vqi+99FL97ne/W0sp9d577+15/IQJE+qWLVtqrbVu3bq11lrrD3/4wzp58uT61FNP1VprTVJvu+22Wmutn/zkJ+tv/MZv1Fprveiii+rf/u3f1h/96Ef17W9/e33yySdrrbUuWrSofuADH+jnV/6TBuo9B4avJKvqIOg7Q+Wvt/4I0NcGcg5o7r1Tb+/57nqk3TL30lvf+tY8+eSTefzxx7Nly5aMHTs2Rx11VH77t387d999dw466KBs2rQpmzdvTpJMmDAhs2fP7nWsBQsWZMmSJUmSDRs25NFHH824ceNy0EEHZd68eUmSX/u1X8uv/uqv7vK8Rx55JN/+9rdz+umnJ0l27NiRo446qr9eMgAAdIS59/4R7vbBeeedl8WLF+d73/tezj///CxcuDBbtmxJV1dXRo4cmYkTJ/YcrvTwww/vdYwVK1bkrrvuyr333ptRo0Zlzpw5r3mI01cfIafWmsmTJ+fee+/t2xcGAACDjLn3vvObu31w/vnnZ9GiRVm8eHHOO++8bNu2LUceeWRGjhyZ5cuXZ/369XscY9u2bRk7dmxGjRqVhx9+OPfdd1/PfS+99FLPkXn++q//Ou985zt3ee5b3vKWbNmypecD9uMf/zgPPvhgH75CAAAYHMy9952Vu30wefLkfP/7388xxxyTo446KhdccEF++Zd/ObNmzcr06dNzwgkn7HGMM888M3/xF3+RqVOn5i1vecsuy8eHH354HnzwwcycOTNjxozJ7bffvstzDznkkCxevDiXX355tm3blu3bt+cjH/lIJk+e3OevFQAAOsnce9+Vnb/JGxpmzZpVV61atctta9euzaRJkzpU0fDkPQf6Wymlq9Y6q9N1DBW99UeAvmYOOPB6e8931yPtlgkAANAA4Q4AAKABfnMHw8jM3/18p0voM13XXdjpEvrMO/7kHZ0uoU+s/PDKTpcAAMOacAcMSf/6h1M6XULfGftTna4AAGiA3TIBAAAaINwBAAA0QLg7QHPmzMmdd965y2033HBDLrvssr0e46yzzsozzzzT16X9hB/+8If5pV/6pZxwwgmZPHlyfu/3fq/nvhdffDHz5s3L8ccfn7e97W1Zt25dv9cDAAB7y7x7z5r7zV1fHzBiTwdtmD9/fhYtWpQzzjij57ZFixbluuuu2+PYtdbUWnPHHXcccJ1763d+53fyrne9Kz/60Y8yd+7cfOlLX8ov/uIv5uabb87YsWPz2GOPZdGiRbnyyit/4kSOAADwMvPu3evEvNvK3QE677zz8sUvfjEvvvhikmTdunV5/PHHM3369MydOzczZszIlClTsnTp0p77J02alMsuuywzZszIhg0bMnHixDz11FNJknPOOSczZ87M5MmTc9NNN/VsZ/To0bn66qszbdq0zJ49O5s3b06SbN68Oeeee26mTZuWadOm5Z577kmS3HbbbTn55JMzffr0fOhDH8qOHTsyatSovOtd70qSHHLIIZkxY0Y2btyYJFm6dGkuuuiinte0bNmyDKUT3AMA0Dbz7j0T7g7QuHHjcvLJJ+fLX/5ykp3fHsybNy+HHXZYlixZktWrV2f58uW54oorev7RHnnkkVx44YW5//77M2HChF3Gu+WWW9LV1ZVVq1ZlwYIF2bp1a5Lkueeey+zZs7NmzZqccsop+exnP5skufzyy3PqqadmzZo1Wb16dSZPnpy1a9fm9ttvz8qVK/PAAw9kxIgRWbhw4S7beeaZZ/IP//APmTt3bpJk06ZNOfbYY5MkBx98cMaMGdOzbQAA6DTz7j1rbrfMTnh5ifhXfuVXsmjRotxyyy2pteaqq67K3XffnYMOOiibNm3qSf0TJkzI7Nmzex1rwYIFWbJkSZJkw4YNefTRRzNu3Lgccsghee9735skmTlzZr7yla8kSb761a/m85/fuSQ+YsSIjBkzJn/1V3+Vrq6unHTSSUmS559/PkceeWTPNrZv35758+fn8ssvz8/+7M8mSa/fFpRS+uLtAQCAPmHevXvCXR8455xz8tGPfjSrV6/O888/nxkzZuTWW2/Nli1b0tXVlZEjR2bixIl54YUXkiSHH354r+OsWLEid911V+69996MGjUqc+bM6XnOyJEje/7RR4wYke3bt79mPbXWXHTRRbn22mt7vf+SSy7Jm9/85nzkIx/puW38+PHZsGFDxo8fn+3bt2fbtm15/etfv1/vBwAA9Afz7t2zW2YfGD16dObMmZMPfvCDmT9/fpJk27ZtOfLIIzNy5MgsX74869ev3+M427Zty9ixYzNq1Kg8/PDDue+++/b4nLlz5+bGG29MkuzYsSPPPvts5s6dm8WLF+fJJ59Mkjz99NM92//4xz+ebdu25YYbbthlnLPPPjuf+9znkiSLFy/Ou9/9bit3AAAMKubduyfc9ZH58+dnzZo1Of/885MkF1xwQVatWpVZs2Zl4cKFOeGEE/Y4xplnnpnt27dn6tSpueaaa15zCfmVPvOZz2T58uWZMmVKZs6cmQcffDAnnnhiPvWpT+U973lPpk6dmtNPPz1PPPFENm7cmD/6oz/KQw89lBkzZmT69On5y7/8yyTJxRdfnK1bt+b444/P9ddfn09/+tMH9oYAAEA/MO9+bWUoHRFx1qxZddWqVbvctnbt2kyaNKlDFQ1P3vOhq68PWdxJS16358MeDxXzx/5Up0voEys/vLLPxiqldNVaZ/XZgI3rrT8C9DVzwIHX23u+ux5p5Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLg7QHPmzMmdd965y2033HBDLrvssr0e46yzzsozzzzT16Xt1tlnn52f//mf77n+4osvZt68eTn++OPztre9LevWrRvQegAAYHfMu/fs4D4baZD41z+c0qfjHfdfvrXb++fPn59FixbljDPO6Llt0aJFue66PZ+Dq9aaWmvuuOOOA65zX/zd3/1dRo8evcttN998c8aOHZvHHnssixYtypVXXpnbb799QOsCAGDoMO/es4Ged1u5O0DnnXdevvjFL+bFF19Mkqxbty6PP/54pk+fnrlz52bGjBmZMmVKli5d2nP/pEmTctlll2XGjBnZsGFDJk6cmKeeeipJcs4552TmzJmZPHlybrrppp7tjB49OldffXWmTZuW2bNnZ/PmzUmSzZs359xzz820adMybdq03HPPPUmS2267LSeffHKmT5+eD33oQ9mxY0eS5Ac/+EGuv/76fPzjH9/ldSxdujQXXXRRz2tatmxZhtIJ7gEAaJt5954Jdwdo3LhxOfnkk/PlL385yc5vD+bNm5fDDjssS5YsyerVq7N8+fJcccUVPf9ojzzySC688MLcf//9mTBhwi7j3XLLLenq6sqqVauyYMGCbN26NUny3HPPZfbs2VmzZk1OOeWUfPazn02SXH755Tn11FOzZs2arF69OpMnT87atWtz++23Z+XKlXnggQcyYsSILFy4MElyzTXX5IorrsioUaN22e6mTZty7LHHJkkOPvjgjBkzpmfbAADQaebdeybc9YGXl4iTnR+y+fPnp9aaq666KlOnTs1pp52WTZs29aT+CRMmZPbs2b2OtWDBgp5vCTZs2JBHH300SXLIIYfkve99b5Jk5syZPfvmfvWrX82ll16aJBkxYkTGjBmTZcuWpaurKyeddFKmT5+eZcuW5Tvf+U4eeOCBPPbYYzn33HN/Yru9fVtQSjmwNwYAAPqQeffuNfebu04455xz8tGPfjSrV6/O888/nxkzZuTWW2/Nli1b0tXVlZEjR2bixIl54YUXkiSHH354r+OsWLEid911V+69996MGjUqc+bM6XnOyJEje/7RR4wYke3bt79mPbXWXHTRRbn22mt3uf3GG29MV1dXJk6cmO3bt+fJJ5/MnDlzsmLFiowfPz4bNmzI+PHjs3379mzbti2vf/3r++LtAQCAPmHevXtW7vrA6NGjM2fOnHzwgx/M/PnzkyTbtm3LkUcemZEjR2b58uVZv379HsfZtm1bxo4dm1GjRuXhhx/Offfdt8fnzJ07NzfeeGOSZMeOHXn22Wczd+7cLF68OE8++WSS5Omnn8769etz6aWX5vHHH8+6devyjW98Iz/3cz+XFStWJNl5FJ/Pfe5zSZLFixfn3e9+t5U7AAAGFfPu3RPu+sj8+fOzZs2anH/++UmSCy64IKtWrcqsWbOycOHCnHDCCXsc48wzz8z27dszderUXHPNNa+5hPxKn/nMZ7J8+fJMmTIlM2fOzIMPPpgTTzwxn/rUp/Ke97wnU6dOzemnn54nnnhit+NcfPHF2bp1a44//vhcf/31+fSnP713LxwAAAaQefdrK0PpiIizZs2qq1at2uW2tWvXZtKkSR2qaHjyng9dM3/3850uoc8sed2eD3s8VMwf+1OdLqFPrPzwyj4bq5TSVWud1WcDNq63/gjQ18wBB15v7/nueqSVOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4e4AzZkzJ3feeecut91www257LLL9nqMs846K88880xfl9arL3zhC5kyZUqmTp2aM888M0899VSS5MUXX8y8efNy/PHH521ve1vWrVs3IPUAAMDeMO/es4P7bKR9VEo5Nsnnk/xMkpeS3FRr/cyBjvuOP3nHgQ6xiz2dt2n+/PlZtGhRzjjjjJ7bFi1alOuu2/M5uGqtqbXmjjvuOOA698b27dvzW7/1W3nooYfyhje8IR/72Mfyp3/6p/nEJz6Rm2++OWPHjs1jjz2WRYsW5corr8ztt98+IHUBADD0mHe/tk7Nuzu5crc9yRW11klJZif5jVLKiR2sZ7+cd955+eIXv5gXX3wxSbJu3bo8/vjjmT59eubOnZsZM2ZkypQpWbp0ac/9kyZNymWXXZYZM2Zkw4YNmThxYk+SP+ecczJz5sxMnjw5N910U892Ro8enauvvjrTpk3L7Nmzs3nz5iTJ5s2bc+6552batGmZNm1a7rnnniTJbbfdlpNPPjnTp0/Phz70oezYsaPnQ/3cc8+l1ppnn302Rx99dJJk6dKlueiii3pe07JlyzKUTnAPAEDbzLv3rGPhrtb6RK11dffl7ydZm+SYTtWzv8aNG5eTTz45X/7yl5Ps/PZg3rx5Oeyww7JkyZKsXr06y5cvzxVXXNHzj/bII4/kwgsvzP33358JEybsMt4tt9ySrq6urFq1KgsWLMjWrVuTJM8991xmz56dNWvW5JRTTslnP/vZJMnll1+eU089NWvWrMnq1aszefLkrF27NrfffntWrlyZBx54ICNGjMjChQszcuTI3HjjjZkyZUqOPvroPPTQQ7n44ouTJJs2bcqxxx6bJDn44IMzZsyYnm0DAECnmXfv2aD4zV0pZWKStyb5p85Wsn9eXiJOdn7I5s+fn1prrrrqqkydOjWnnXZaNm3a1JP6J0yYkNmzZ/c61oIFC3q+JdiwYUMeffTRJMkhhxyS9773vUmSmTNn9uyb+9WvfjWXXnppkmTEiBEZM2ZMli1blq6urpx00kmZPn16li1blu985zv58Y9/nBtvvDH3339/Hn/88UydOjXXXnttkvT6bUEppe/eJAAAOEDm3bvXsd/cvayUMjrJ/0zykVrrs73cf0mSS5LkuOOOG+Dq9s4555yTj370o1m9enWef/75zJgxI7feemu2bNmSrq6ujBw5MhMnTswLL7yQJDn88MN7HWfFihW56667cu+992bUqFGZM2dOz3NGjhzZ848+YsSIbN++/TXrqbXmoosu6vkAveyb3/xmkuRNb3pTkuR973tfPv3pTydJxo8fnw0bNmT8+PHZvn17tm3blte//vUH8K4A0J+GQn8E6Gvm3bvX0ZW7UsrI7Ax2C2utf9fbY2qtN9VaZ9VaZx1xxBEDW+BeGj16dObMmZMPfvCDmT9/fpJk27ZtOfLIIzNy5MgsX74869ev3+M427Zty9ixYzNq1Kg8/PDDue+++/b4nLlz5+bGG29MkuzYsSPPPvts5s6dm8WLF+fJJ59Mkjz99NNZv359jjnmmDz00EPZsmVLkuQrX/lKJk2alCQ5++yz87nPfS5Jsnjx4rz73e+2cgcwiA2F/gjQ18y7d69j4a7sfAU3J1lba72+U3X0lfnz52fNmjU5//zzkyQXXHBBVq1alVmzZmXhwoU54YQT9jjGmWeeme3bt2fq1Km55pprXnMJ+ZU+85nPZPny5ZkyZUpmzpyZBx98MCeeeGI+9alP5T3veU+mTp2a008/PU888USOPvro/MEf/EFOOeWUTJ06NQ888ECuuuqqJMnFF1+crVu35vjjj8/111/f880CAAAMJubdr6106oiIpZR3Jvl6km9l56kQkuSqWutrHp901qxZddWqVbvctnbt2p4UzMDwng9dM3/3850uoc8sed2eD3s8VMwf+1OdLqFP7OkQ1vuilNJVa53VZwM2rrf+CNDXzAEHXm/v+e56ZMd+c1dr/UYS+/0BAAD0gUFxtEwAAAAOjHAHAADQgCbCXad+Nzgcea8BAIYvc8GBsz/v9ZAPd4ceemi2bt3qgzYAaq3ZunVrDj300E6XAgDAADPvHjj7O+/u+EnMD9T48eOzcePGnnNI0L8OPfTQjB8/vtNlAAAwwMy7B9b+zLuHfLgbOXJk3vjGN3a6DAAAaJp59+A35HfLBAAAQLgDAABognAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAAzoa7kopt5RSniylfLuTdQAAAAx1nV65uzXJmR2uAQAAYMjraLirtd6d5OlO1gAAANCCTq/cAQAA0AcGfbgrpVxSSllVSlm1ZcuWTpcDAIOC/gjAqw36cFdrvanWOqvWOuuII47odDkAMCjojwC82qAPdwAAAOxZp0+F8IUk9yZ5SyllYynl4k7WAwAAMFQd3MmN11rnd3L7AAAArbBbJgAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAbsMdyVUv5jKeXmUsqXuq+fWEq5uP9LA4DhSe8FYH/szcrdrUnuTHJ09/V/SfKR/ioIANB7Adh3exPu3lBr/ZskLyVJrXV7kh39WhUADG96LwD7bG/C3XOllHFJapKUUmYn2davVQHA8Kb3ArDPDt6Lx3w0yd8neVMpZWWSI5Kc169VAcDwpvcCsM/2GO5qratLKacmeUuSkuSRWuuP+70yABim9F4A9scew10p5cJX3TSjlJJa6+f7qSYAGNb0XgD2x97slnnSKy4fmmRuktVJNBgA6B96LwD7bG92y/zwK6+XUsYk+at+qwgAhjm9F4D9sTdHy3y1HyZ5c18XAgC8Jr0XgD3am9/c/UO6D8WcnWHwxCR/059FAcBwpvcCsD/25jd3//0Vl7cnWV9r3dhP9QAAei8A+2FvfnP3tYEoBADYSe8FYH+8ZrgrpXw//75LyC53Jam11p/qt6oAYBjSewE4EK8Z7mqtrxvIQgBguNN7ATgQe/ObuyRJKeXI7DzXTpKk1vqv/VIRAJBE7wVg3+zxVAillLNLKY8m+W6SryVZl+RL/VwXAAxbei8A+2NvznP3X5PMTvIvtdY3JpmbZGW/VgUAw5veC8A+25tw9+Na69YkB5VSDqq1Lk8yvZ/rAoDhTO8FYJ/tzW/unimljE7y9SQLSylPZuc5dwCA/qH3ArDP9mbl7u4kP53kt5J8Ocn/l+SX+7MoABjm9F4A9tnehLuS5M4kK5KMTnJ7964iAED/0HsB2Gd7DHe11k/WWicn+Y0kRyf5Winlrn6vDACGKb0XgP2xNyt3L3syyTaFs+sAAAvkSURBVPeSbE1yZP+UAwC8gt4LwF7bm/PcXVpKWZFkWZI3JPn1WuvU/i4MAIYrvReA/bE3R8uckOQjtdYH+rsYACCJ3gvAfthjuKu1/t5AFAIA7KT3ArA/9uU3dwAAAAxSwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0oKPhrpRyZinlkVLKY6UUh30GAADYTx0Ld6WUEUn+LMkvJjkxyfxSyomdqgcAAGAo6+TK3clJHqu1fqfW+qMki5L8SgfrAQAAGLI6Ge6OSbLhFdc3dt8GAADAPjq4g9suvdxWf+JBpVyS5JIkOe644/ps4zN/9/N9NlanLXnddZ0uoU/MH/tTnS6hz6z88MpOl9Crrusu7HQJfaid1zI4Py0Mdv3VH5N2emQr/THRIweCz/3g43O/7zq5crcxybGvuD4+yeOvflCt9aZa66xa66wjjjhiwIoDgMFMfwTg1ToZ7r6Z5M2llDeWUg5Jcn6Sv+9gPQAAAENWx3bLrLVuL6X8ZpI7k4xIckut9cFO1QMAADCUdfI3d6m13pHkjk7WAAAA0IKOnsQcAACAviHcAQAANKCju2UCAMBg0M7pglp5HUn+5B2drmDIsXIHAADQACt3AADAoDNYT3g/mFm5AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAQd3ugAAYHDpuu7CTpfQR1p5HcnKThcADAlW7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0oCPhrpTyf5VSHiylvFRKmdWJGgAAAFrSqZW7byf51SR3d2j7AAAATTm4Exutta5NklJKJzYPAADQHL+5AwAAaEC/rdyVUu5K8jO93HV1rXXpPoxzSZJLkuS4447ro+oAYGjTHwF4tX4Ld7XW0/ponJuS3JQks2bNqn0xJgAMdfojAK9mt0wAAIAGdOpUCOeWUjYmeXuS/11KubMTdQAAALSiU0fLXJJkSSe2DQAA0CK7ZQIAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABog3AEAADRAuAMAAGiAcAcAANAA4Q4AAKABwh0AAEADhDsAAIAGCHcAAAANEO4AAAAaINwBAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADRDuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA0Q7gAAABpwcKcLgJet/PDKTpcAAABDlpU7AACABli5a8Bx/+VbnS4BAADosGEb7rquu7DTJQAAAPQZu2UCAAA0QLgDAABogHAHAADQAOEOAACgAcIdAABAA4Q7AACABgh3AAAADehIuCulXFdKebiU8s+llCWllJ/uRB0AAACt6NTK3VeS/HytdWqSf0ny+x2qAwAAoAkdCXe11n+stW7vvnpfkvGdqAMAAKAVg+E3dx9M8qVOFwEAADCUHdxfA5dS7kryM73cdXWtdWn3Y65Osj3Jwt2Mc0mSS5LkuOOO64dKAWDo0R8BeLV+C3e11tN2d38p5aIk700yt9ZadzPOTUluSpJZs2a95uMAYDjRHwF4tX4Ld7tTSjkzyZVJTq21/rATNQAAALSkU7+5+9Mkr0vylVLKA6WUv+hQHQAAAE3oyMpdrfX4TmwXAACgVYPhaJkAAAAcIOEOAACgAcIdAABAA8puzkIw6JRStiRZ3+k6hqk3JHmq00XAAPO575wJtdYjOl3EUKE/dpz/VjAc+dx3zmv2yCEV7uicUsqqWuusTtcBA8nnHtgb/lvBcORzPzjZLRMAAKABwh0AAEADhDv21k2dLgA6wOce2Bv+W8Fw5HM/CPnNHQAAQAOs3AEAADRAuKNHKeUHna4BBkIpZWIp5du93P6HpZTT9vDcT5RSfqf/qgMGG/2R4USPHNoO7nQBAINFrfW/dLoGABiM9MihwcodP6HsdF0p5dullG+VUuZ13/7npZSzuy8vKaXc0n354lLKpzpZM+yHEaWUz5ZSHiyl/GMp5bBSyq2llPOSpJRyVinl4VLKN0opC0opX3zFc08spawopXynlHJ5h+oHBpj+yDCiRw5Rwh29+dUk05NMS3JakutKKUcluTvJL3Q/5pgkJ3ZffmeSrw90kXCA3pzkz2qtk5M8k+Q/v3xHKeXQJP8jyS/WWt+Z5IhXPfeEJGckOTnJH5RSRg5MyUCH6Y8MF3rkECXc0Zt3JvlCrXVHrXVzkq8lOSk7G9QvlFJOTPJQks3dTe3tSe7pWLWwf75ba32g+3JXkomvuO+EJN+ptX63+/oXXvXc/11rfbHW+lSSJ5P8x36tFBgs9EeGCz1yiPKbO3pTerux1rqplDI2yZnZ+S3l65O8L8kPaq3fH8D6oC+8+IrLO5Ic9orrvf5/YDfP9d9SGB70R4YLPXKIsnJHb+5OMq+UMqKUckSSU5L8n+777k3yke7HfD3J78QuJ7Tn4SQ/W0qZ2H19XudKAQYR/RH0yEFNkqY3S7JzV5I1SWqSj9Vav9d939eTvKfW+lgpZX12fjupedGUWuvzpZTLkny5lPJU/n3yBgxv+iPDnh45uJVaa6drABh0Simja60/KKWUJH+W5NFa6//T6boAoNP0yMHLbpkAvfv1UsoDSR5MMiY7jwwGAOiRg5aVOwAAgAZYuQMAAGiAcAcAANAA4Q4AAKABwh0MYaWUL5RS/rmU8tudrgUABhM9kuHIAVVgiCql/EySf6q1Tuh0LQAwmOiRDFdW7qCDSikTSykPl1L+spTy7VLKwlLKaaWUlaWUR0spJ3f/3VNKub/7f9/S/fR/THJkKeWBUsovdPJ1AEBf0yNh31m5gw4qpUxM8liSt2bnuWK+mWRNkouTnJ3kA0kuTPLDWuv2UsppSS6ttf7n7ud+sdb68wNfOQD0Lz0S9t3BnS4AyHdrrd9KklLKg0mW1VprKeVbSSZm58lBP1dKeXOSmmRkxyoFgIGlR8I+sFsmdN6Lr7j80iuuv5SdX8D81yTLu799/OUkhw5seQDQMXok7APhDga/MUk2dV9+fwfrAIDBRo+EVxDuYPD74yTXllJWJhnR6WIAYBDRI+EVHFAFAACgAVbuAAAAGiDcAQAANEC4AwAAaIBwBwAA0ADhDgAAoAHCHQAAQAOEOwAAgAYIdwAAAA34/wEI0PPsqhUDawAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1080x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "low = hsq_var_melt[hsq_var_melt.ld=='low']\n",
    "high = hsq_var_melt[hsq_var_melt.ld=='high']\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(ncols=2, sharey=True)\n",
    "fig.set_figheight(8)\n",
    "fig.set_figwidth(15)\n",
    "sns.barplot(low.maf, low.value, hue=low.variable, ax=ax1)\n",
    "ax1.title.set_text(\"Low LD\")\n",
    "\n",
    "sns.barplot(high.maf, high.value, hue=high.variable, ax=ax2)\n",
    "ax2.title.set_text(\"High LD\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.7",
   "language": "python",
   "name": "py3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
